vcd/resource_vcd_edgegateway.go

```diff
@@ -0,0 +1,201 @@
+				Default:     false,
+				ForceNew:    true,
+				Description: "True if the gateway uses advanced networking",
+			},
+			"backing_config": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Configuration of the vShield edge VM for this gateway. One of: compact, full",
+			},
+			"ha_enabled": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				ForceNew:    true,
+				Description: "Enable high availability on this edge gateway",
+			},
+			"external_networks": &schema.Schema{
+				Type:        schema.TypeList,
+				Required:    true,
+				ForceNew:    true,
+				Description: "A list of external networks to be used by the edge gateway",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"default_gateway": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "",
+				ForceNew:    true,
+				Description: "External network to be used as default gateway. Its name must be included in 'external_networks'. An empty value will skip the default gateway",
+			},
+			"distributed_routing": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				ForceNew:    true,
+				Description: "If advanced networking enabled, also enable distributed routing",
+			},
+		},
+	}
+}
+
+// Creates a new edge gateway from a resource definition
+func resourceVcdEdgeGatewayCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[TRACE] edge gateway creation initiated")
+
+	vcdClient := meta.(*VCDClient)
+
+	rawExternalNetworks := d.Get("external_networks").([]interface{})
+	var externalNetworks []string
+	for _, en := range rawExternalNetworks {
+		externalNetworks = append(externalNetworks, en.(string))
+	}
+
+	var gwCreation = govcd.EdgeGatewayCreation{
+		ExternalNetworks:          externalNetworks,
+		Name:                      d.Get("name").(string),
+		OrgName:                   d.Get("org").(string),
+		VdcName:                   d.Get("vdc").(string),
+		Description:               d.Get("description").(string),
+		BackingConfiguration:      d.Get("backing_config").(string),
+		AdvancedNetworkingEnabled: d.Get("advanced_networking").(bool),
+		DefaultGateway:            d.Get("default_gateway").(string),
+		DistributedRoutingEnabled: d.Get("distributed_routing").(bool),
+		HAEnabled:                 d.Get("ha_enabled").(bool),
+	}
+
+	edge, err := govcd.CreateEdgeGateway(vcdClient.VCDClient, gwCreation)
+	if err != nil {
+		log.Printf("[DEBUG] Error creating edge gateway: %#v", err)
+		return fmt.Errorf("error creating edge gateway: %#v", err)
+	}
+
+	d.SetId(d.Get("name").(string))
+	log.Printf("[TRACE] edge gateway created: %#v", edge.EdgeGateway.Name)
+	return resourceVcdEdgeGatewayRead(d, meta)
+}
+
+func setEdgeGatewayValues(d *schema.ResourceData, egw govcd.EdgeGateway) error {
+
+	d.SetId(egw.EdgeGateway.HREF)
```

Comment:
Here it is `id=https://192.168.1.109/api/admin/edgeGateway/f2c007f5-ac52-4f57-8a57-b6a1f261a875`
I'd vote for going to real edge gateway ID (not name and not full url)

Reply:
Changed to using ID. Note, however, that we don't have any code that uses the ID, and we need to make changes to start using it in data sources.
We need to clean up the code and set expectations about what we will find in those resource IDs when we consume them. So far, there are a few cases where we expect to find the resource name, but we should change that.

Reply:
Yes. I know, but one can always refer to `.name` of resource for the time being I guess (at least for new resources as the edge gateway is).

Whoever use this in future on other resources we can add `edge_gateway_id` attribute instead of `edge_gateway` which expects name now. I think this is consistent now with our idea of moving towards real IDs.

I think I like it now (`id=urn:vcloud:gateway:d87f86b3-bc79-4322-93b5-91f6f9ee37af`) although I had thoughts about storing UUID only (`d87f86b3-bc79-4322-93b5-91f6f9ee37af`) however no one prevents to strip the ID later.